core/common/src/main/scala/org/typelevel/otel4s/context/Context.scala

```diff
@@ -0,0 +1,76 @@
+/*
+ * Copyright 2022 Typelevel
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.typelevel.otel4s
+package context
+
+trait Context[C] { outer =>
```

Comment:
According to the signature, the `Context` here acts like a delegate or proxy instance.

If I get it right, the `Context` is a key-value storage. While we delegate things to the `JContext` under the hood, it's not important for the parent interface.

Can we define the context as follows?
```scala
trait Context {
  type Key[A] <: context.Key[A]
  type Self <: Context
  
  def get[A](key: Key[A]): Option[A]
  def getOrElse[A](key: Key[A], default: => A): A = get(key).getOrElse(default)
  def updated[A](key: Key[A], value: A): Self 
}
```

Then, we don't need `ContextOps`.

Reply:
> Can we define the context as follows?

no :( this is "unsound" and not permitted by scala 3 or native.

here's my unpolished first design which is along the same lines: https://github.com/typelevel/otel4s/commit/dc46bd45d5d3a2f27407bf341777daaf76cd3e88

Reply:
I agree that the design you describe is SO much nicer. I think it's easier to implement, easier to use, simpler, and more intuitive. sadly, there's nothing requiring that `Context#Key` and `Context#Self` be the same for all instances of a given `C <: Context`, so even though this design uses it soundly, the compiler can't guarantee that